### PR TITLE
Feat (stats): Statistics capture

### PR DESCRIPTION
**Pull request by pablomlago**
## Reason for this PR

Add opt-in statistics collection for GPTQ/PTQ to quantify and debug per-layer quantization error during calibration, without changing default behavior.

### What
- Introduced a generic statistics collection framework (`stats_utils.py`) with a no-op default.
- Instrumented GPTQ to emit `pre_update` and `post_update` events, to compute statistics before/after weight update.
- Added GPTQ error metrics (relative weight error and Hessian-weighted output error).
- Added `--ptq-stats` CLI flag to the LLM entrypoint.

### How
- Used a `ContextVar`-based collector to minimize changes required for statistics collection.
- Decoupled *when* stats are emitted (events) from *what* is computed (handlers).

## Testing Summary

- Manually validated GPTQ runs with and without `--ptq-stats gptq`.

## Risk Highlight

- [ ] This PR includes code from another work (please detail).
- [ ] This PR contains API-breaking changes.
- [ ] This PR depends on work in another PR (please provide links/details).
- [ ] This PR introduces new dependencies (please detail).
- [ ] There are coverage gaps not covered by tests.
- [ ] Documentation updates required in subsequent PR.

## Checklist

- [ ] Code comments added to any hard-to-understand areas, if applicable.
- [ ] Changes generate no new warnings.
- [ ] Updated any relevant tests, if applicable.
- [ ] No conflicts with destination `dev` branch.
- [ ] I reviewed my own code changes.
- [ ] Initial CI/CD passing.
- [ ] 1+ reviews given, and any review issues addressed and approved.
- [ ] Post-review full CI/CD passing.

### CODE REVIEW

File: src/brevitas_examples/llm/main.py [hunk 1]
```diff
@@ -607,16 +609,19 @@ def quantize_llm(args, extra_args=None):
...
-                max_accumulator_tile_size=args.gpxq_max_accumulator_tile_size)
+            stats_collector_ctx = collect_stats(
+                DictStatsCollector()) if "gptq" in args.ptq_stats else nullcontext()
+            with stats_collector_ctx:
+                apply_gptq(
+                    model,
+                    calibration_loader,
+                    act_order=args.gpxq_act_order,
+                    use_quant_activations=args.gpxq_use_quant_activations,
+                    create_weight_orig=not args.disable_create_weight_orig,
```
Comment (JP-Amboage) [inline src/brevitas_examples/llm/main.py:620]: Given the current code, I think this has to be true anytime the stats are being logged. I'd suggest adding a validation for that.

File: src/brevitas/utils/stats_utils.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+class BaseStatsCollector(ABC):
+
+    def __init__(self) -> None:
+        self._stats_fn: Dict[str, StatFn] = {}
+
+    def on(self, key: str, fn: StatFn) -> None:
+        self._stats_fn[key] = fn
+
+    def log(self, event: str, **payload) -> None:
+        if (fn := self._stats_fn.get(event, None)) is not None:
```
Comment (JP-Amboage) [inline src/brevitas/utils/stats_utils.py]: Very minor thing, but `None`  is already the default value in `.get` so this could just be `self._stats_fn.get(event)`. Also, since the values of `self._stats_fn` are always `StatFn`, I think we can avoid the `is not None`. And just rewrite the whole line as `if (fn := self._stats_fn.get(event))`. But this is just a matter of "style" so if you find the current one to be more readable or prefer it for any other reason I'm fine with that too.

File: src/brevitas/graph/utils.py [hunk 3]
```diff
@@ -186,3 +186,22 @@ def remove_weight_orig(model: nn.Module):
     for name, module in model.named_modules():
         if hasattr(module, 'weight_orig'):
             del module.weight_orig
+
+
+@torch.no_grad()
+def gpxq_compute_error_stats(prefix: str, name: str, layer: torch.nn.Module,
+                             H: torch.Tensor) -> Dict[str, float]:
+    # This computation only supports nn.Linear, currently
+    assert isinstance(layer, qnn.QuantLinear), "`compute_weight_error` only supports `QuantLinear` layers"
```
Comment (JP-Amboage) [inline src/brevitas/graph/utils.py]: Maybe, following our private discussion, replace `assert` for `Exceptions/Errors`

File: src/brevitas/graph/utils.py [hunk 4]
```diff
@@ -186,3 +186,22 @@ def remove_weight_orig(model: nn.Module):
...
+    quant_weight = layer.quant_weight().value.to(dtype=H.dtype)
+    weight = layer.weight_orig.to(dtype=H.dtype, device=quant_weight.device)
+    H = H.to(device=quant_weight.device).squeeze(0)
```
Comment (JP-Amboage) [inline src/brevitas/graph/utils.py]: perhaps the device could be a parameter to accommodate for cases where memory is an issue. For example, in cases where it is possible to run the whole algorithm (GPTQ/Qronos) in GPU memory, but there's no space for these extra weights or H matrix in GPU. In the current approach, one would have to switch the whole thing to CPU, however it could be better to still run the algorithm in GPU and just the stats in CPU.

File: src/brevitas/graph/gptq.py [hunk 5]
```diff
@@ -112,6 +119,13 @@ def single_layer_update(self, percdamp=.01, c=1e4):
         weight = self.layer.weight.data
         dev = weight.device
 
+        gptq_stats_collector = StatsCollectorCtx.get()
+        H_orig = None
+        if gptq_stats_collector.is_active:
+            H_orig = deepcopy(self.H)
```
Comment (JP-Amboage) [inline src/brevitas/graph/gptq.py]: As I also mentioned in stats_utils.py, perhaps allow for using an specific device for the copy. I explained why I think this in the other comment (stats_utils.py).

File: src/brevitas/graph/utils.py [hunk 6]
```diff
@@ -186,3 +186,22 @@ def remove_weight_orig(model: nn.Module):
     for name, module in model.named_modules():
         if hasattr(module, 'weight_orig'):
             del module.weight_orig
+
+
+@torch.no_grad()
+def gpxq_compute_error_stats(prefix: str, name: str, layer: torch.nn.Module,
```
Comment (JP-Amboage) [inline src/brevitas/graph/utils.py]: Since this is for `gpxq` It may be good to have also an argument to receive `G` (the "cross correlation" matrix in Qronos) defaulting to None. Then inside the function decide which loss to compute based on `G` being None or not. This would allow this function to be directly applicable in Qronos to compute stats about the Qronos objective.

File: src/brevitas/graph/gptq.py [hunk 7]
```diff
@@ -261,6 +284,13 @@ def __init__(
         self.num_blocks = num_blocks
         self.gptq_class = gptq_class
 
+        # Register the statistics to collect during GPTQ
+        gptq_stats_collector = StatsCollectorCtx.get()
+        gptq_stats_collector.on(
+            "pre_update", functools.partial(gpxq_compute_error_stats, prefix="pre"))
+        gptq_stats_collector.on(
+            "post_update", functools.partial(gpxq_compute_error_stats, prefix="post"))
```
Comment (JP-Amboage) [inline src/brevitas/graph/gptq.py]: I wonder if this could be in the parent GPXQ class to avoid duplicating code when extending this functionality to Qronos etc.

File: src/brevitas/graph/gptq.py [hunk 8]
```diff
@@ -195,6 +209,15 @@ def single_layer_update(self, percdamp=.01, c=1e4):
                 weight[group_index, :, perm[i2:]] -= (
                     error_block[group_index].matmul(h_inv[group_index, i1:i2,
                                                           i2:].to(dev))).to(dtype)
+
+        # Log post-update statistics
+        gptq_stats_collector.log("post_update", name=self.name, layer=self.layer, H=H_orig)
```
Comment (JP-Amboage) [inline src/brevitas/graph/gptq.py]: I wonder if this logic (and other logging related logic) could be encapsulated in a function/method and handled through the parent class so that the functionality extends directly to other GPXQ methods.

File: src/brevitas_examples/llm/llm_args.py [hunk 9]
```diff
@@ -486,6 +486,13 @@ def create_args_parser() -> ArgumentParser:
         default=1,
         help='Batch size for calibration data loader. (default: %(default)s).')
 
+    parser.add_argument(
+        '--ptq-stats',
```
Comment (JP-Amboage) [inline src/brevitas_examples/llm/llm_args.py:490]: Is the use of PTQ here intentional (to keep it more general)? Or should it be GPXQ?

File: src/brevitas/graph/gpxq.py [hunk 10]
```diff
@@ -110,6 +117,13 @@ def __init__(
         else:
             self.model.forward = self.catch_stopfwd
 
+        # Register the statistics to collect during GPTQ
+        gptq_stats_collector = StatsCollectorCtx.get()
+        gptq_stats_collector.on(
+            "pre_update", functools.partial(gpxq_compute_error_stats, prefix="pre"))
+        gptq_stats_collector.on(
+            "post_update", functools.partial(gpxq_compute_error_stats, prefix="post"))
```
Comment (JP-Amboage) [inline src/brevitas/graph/gpxq.py:125]: perhaps `gpxq_stats_collector` is a better name than `gptq_stats_collector` now that it's shared among GPXQ algorithms.

File: src/brevitas/graph/utils.py [hunk 11]
```diff
@@ -189,16 +191,26 @@ def remove_weight_orig(model: nn.Module):
...
 @torch.no_grad()
-def gpxq_compute_error_stats(prefix: str, name: str, layer: torch.nn.Module,
-                             H: torch.Tensor) -> Dict[str, float]:
+def gpxq_compute_error_stats(
+        prefix: str,
+        name: str,
+        layer: torch.nn.Module,
+        H: torch.Tensor,
+        G: Optional[torch.Tensor] = None,
+        device: Optional[Union[str, torch.device]] = None) -> Dict[str, float]:
```
Comment (JP-Amboage) [inline src/brevitas/graph/utils.py:200]: shall we have a new argument for the llm quantization script like `--gpxq-stats-device` to manage this? I'm not sure if it's better to add that now or to just leave it with the default that you wrote, and let anyone add it in the future if they actually need it.

**Review by JP-Amboage — COMMENTED** [review]
I like the new approach as a solution to get the stats functionality to work in all gpxq algorithms without any extra code 🚀!  I just left two comments, one about a variable name, and another about the device handling. I'm closing the comments from the previous review that have been addressed. Some remain open.

File: src/brevitas/graph/utils.py (hunk 3 above)
Comment (JP-Amboage) [reply src/brevitas/graph/utils.py]: Addressed in the next commit.

File: src/brevitas/graph/utils.py (hunk 4 above)
Comment (JP-Amboage) [reply src/brevitas/graph/utils.py]: Addressed in the next commit

File: src/brevitas/graph/gptq.py (hunk 5 above)
Comment (JP-Amboage) [reply src/brevitas/graph/gptq.py]: Addressed in the next commit

File: src/brevitas/graph/utils.py (hunk 6 above)
Comment (JP-Amboage) [reply src/brevitas/graph/utils.py]: Addressed in the next commit

File: src/brevitas/graph/gptq.py (hunk 7 above)
Comment (JP-Amboage) [reply src/brevitas/graph/gptq.py]: Addressed in the next commit

File: src/brevitas_examples/llm/main.py (hunk 1 above)
Comment (pablomlago) [reply src/brevitas_examples/llm/main.py:620]: Added the following check under the `__init__` of `gpxq_mode`:
```
# Verify that the original weights are kept when statistics are being collected
if gpxq_stats_collector.is_active and not create_weight_orig:
    raise ValueError(
        "Statistics collection during GPxQ requires keeping original weights. Set `create_weight_orig` to True."
    )
```

File: src/brevitas_examples/llm/llm_args.py (hunk 9 above)
Comment (pablomlago) [reply src/brevitas_examples/llm/llm_args.py:490]: Yes, I want to keep it to general, as this is likely to be reused for other PTQ algorithms.

File: src/brevitas/graph/utils.py (hunk 11 above)
Comment (pablomlago) [reply src/brevitas/graph/utils.py:200]: Definitely computing the statistics will increase the peak memory usage during GPxQ, but since I am expecting the statistics collection to be used sporadically, I'd rather delay adding a new CLI argument, until it is strictly needed (e.g. when hitting an OOM for a model that we need to collect debugging statistics on while running GPxQ). In the meantime, I opted for providing the minimal functionality to propagate `device` to the quantization error computation.

File: src/brevitas_examples/llm/main.py (hunk 1 above)
Comment (JP-Amboage) [reply src/brevitas_examples/llm/main.py:620]: Following our private discussion, this check should be performed in `src/brevitas_examples/llm
/llm_args.py`
